csi/controller.go

```diff
@@ -315,11 +316,152 @@ func osdVolumeAttributes(v *api.Volume) map[string]string {
+
+	// Check if the volume has already been created or is in process of creation
+	v, err := util.VolumeFromName(s.driver, req.GetName())
+	if err == nil {
+		osdToCsiVolumeInfo(volume, v)
+		return resp, nil
+	}
+
+	// Get parameters
+	spec, locator, source, err := s.specHandler.SpecFromOpts(req.GetParameters())
+	if err != nil {
+		e := fmt.Sprintf("Unable to get parameters: %s\n", err.Error())
+		dlog.Errorln(e)
+		return nil, status.Error(codes.Internal, e)
+	}
+
+	// Check if the caller is asking to create a snapshot or for a new volume
+	var id string
+	if source != nil && len(source.GetParent()) != 0 {
+		// Get parent volume information
+		parent, err := util.VolumeFromName(s.driver, source.Parent)
+		if err != nil {
+			e := fmt.Sprintf("unable to get parent volume information: %s\n", err.Error())
+			dlog.Errorln(e)
+			return nil, status.Error(codes.InvalidArgument, e)
+		}
+
+		// Create a snapshot from the parent
+		id, err = s.driver.Snapshot(parent.GetId(), false, &api.VolumeLocator{
+			Name: req.GetName(),
+		})
+		if err != nil {
+			e := fmt.Sprintf("unable to create snapshot: %s\n", err.Error())
+			dlog.Errorln(e)
+			return nil, status.Error(codes.Internal, e)
+		}
+	} else {
+		// Get Capabilities and Size
+		spec.Size = req.GetCapacityRange().GetRequiredBytes()
+		for _, cap := range req.GetVolumeCapabilities() {
+			// Check access mode is setup correctly
+			mode := cap.GetAccessMode().GetMode()
+			if mode == csi.VolumeCapability_AccessMode_MULTI_NODE_MULTI_WRITER ||
+				mode == csi.VolumeCapability_AccessMode_MULTI_NODE_SINGLE_WRITER {
+				spec.Shared = true
+			}
+		}
+
+		// Create the volume
+		locator.Name = req.GetName()
+		id, err = s.driver.Create(locator, source, spec)
+		if err != nil {
+			return nil, status.Error(codes.Internal, err.Error())
+		}
+	}
+
+	// id must have been set
+	v, err = util.VolumeFromName(s.driver, id)
+	if err != nil {
+		e := fmt.Sprintf("Unable to find newly created volume: %s", err.Error())
+		dlog.Errorln(e)
+		return nil, status.Error(codes.Internal, e)
+	}
+	osdToCsiVolumeInfo(volume, v)
+	return resp, nil
+
+}
+
+// DeleteVolume is a CSI API which deletes a volume
+func (s *OsdCsiServer) DeleteVolume(
+	ctx context.Context,
+	req *csi.DeleteVolumeRequest,
+) (*csi.DeleteVolumeResponse, error) {
+
+	// Log request
+	dlog.Debugf("DeleteVolume req[%#v]", *req)
+
+	// Check arguments
+	if req.GetVersion() == nil {
+		return nil, status.Error(codes.InvalidArgument, "Version must be provided")
+	}
+	if len(req.GetVolumeId()) == 0 {
+		return nil, status.Error(codes.InvalidArgument, "Volume id must be provided")
+	}
+
+	err := s.driver.Delete(req.GetVolumeId())
+	if err != nil {
+		e := fmt.Sprintf("Unable to delete volume with id %s: %s",
+			req.GetVolumeId(),
+			err.Error())
+		dlog.Errorln(e)
+		return nil, status.Error(codes.Internal, e)
+	}
+
+	return &csi.DeleteVolumeResponse{
+		Reply: &csi.DeleteVolumeResponse_Result_{
+			Result: &csi.DeleteVolumeResponse_Result{},
+		},
```

Comment:
is the response supposed to be empty?

Reply:
Yeah, according to the spec. I found that really weird also :)